yt_dlp/networking/_requests.py

```diff
@@ -0,0 +1,420 @@
+        session.mount('http://', _http_adapter)
+        session.cookies = cookiejar
+        session.trust_env = False  # no need, we already load proxies from env
+        return session
+
+    def _send(self, request):
+
+        headers = self._merge_headers(request.headers)
+        add_accept_encoding_header(headers, SUPPORTED_ENCODINGS)
+
+        max_redirects_exceeded = False
+
+        session = self._get_instance(
+            cookiejar=request.extensions.get('cookiejar') or self.cookiejar
+        )
+
+        try:
+            res = session.request(
+                method=request.method,
+                url=request.url,
+                data=request.data,
+                headers=headers,
+                timeout=float(request.extensions.get('timeout') or self.timeout),
+                proxies=request.proxies or self.proxies,
+                allow_redirects=True,  # TODO(future): add extension to enable/disable redirects
+                stream=True
+            )
+
+        except requests.exceptions.TooManyRedirects as e:
+            max_redirects_exceeded = True
+            res = e.response
+        except requests.exceptions.SSLError as e:
+            if 'CERTIFICATE_VERIFY_FAILED' in str(e):
+                raise CertificateVerifyError(cause=e) from e
+            raise SSLError(cause=e) from e
+        except requests.exceptions.ProxyError as e:
+            raise ProxyError(cause=e) from e
+        except (requests.exceptions.ConnectionError, requests.exceptions.Timeout) as e:
+            # Some urllib3 exceptions such as IncompleteRead are wrapped by ConnectionError on request
+            handle_urllib3_read_exceptions(find_original_error(e, (urllib3.exceptions.HTTPError,)))
+            raise TransportError(cause=e) from e
+        except urllib3.exceptions.HTTPError as e:
+            # Catch any urllib3 exceptions that may leak through
+            handle_urllib3_read_exceptions(e)
+            raise TransportError(cause=e) from e
+        # Any misc Requests exception. May not necessary be network related e.g. InvalidURL
+        except requests.exceptions.RequestException as e:
+            raise RequestError(cause=e) from e
+
+        requests_res = RequestsResponse(res)
+
+        if not 200 <= requests_res.status < 300:
+            raise HTTPError(requests_res, redirect_loop=max_redirects_exceeded)
+
+        return requests_res
+
+
+# Since we already have a socks proxy implementation,
+# we can use that with urllib3 instead of requiring an extra dependency.
+class SocksHTTPConnection(urllib3.connection.HTTPConnection):
+    def __init__(self, _socks_options, *args, **kwargs):  # must use _socks_options to pass PoolKey checks
+        self._proxy_args = _socks_options
+        super().__init__(*args, **kwargs)
+
+    def _new_conn(self):
+        sock = sockssocket()
+        sock.setproxy(**self._proxy_args)
+        if isinstance(self.timeout, (int, float)):
+            sock.settimeout(self.timeout)
```

Comment:
missing source_address setting.

should probably copy something similar to urllib _create_connection